### PR TITLE
LADSPA effect unit refactor

### PR DESCRIPTION
**Pull request by mawe42**
This pull request contains a major refactor of the LADSPA subsystem and the associated shell commands. In it's current form it is more of an RFC than a proper pull request. I'd really like to get your feedback on the size and "mergeability" of this change.

I've basically completely rewritten src/bindings/fluid_ladspa.c. Also, the shell commands have been streamlined, removing the need for extremely long command lines. So something like this:

    ladspa_add ./aw.so alienwah_stereo Input1 <- in1_L Input2 <- in1_R Output1 -> out1_L Output2 -> out2_R Frequency <- $0.6 Initial <- $0 Feedback <- $0.5 Delay <- $20

becomes

    ladspa_plugin ./aw.so alienwah_stereo
    ladspa_port Input1 < in1_L
    ladspa_port Input2 < in1_R
    ladspa_port Output1 > out1_L
    ladspa_port Output2 > out1_R
    ladspa_port Frequency = 0.6
    ladspa_port Initial = 0
    ladspa_port Feedback = 0.5
    ladspa_port Delay = 2.0

I don't think we need to be careful of backwards compatibility here, as the LADSPA system was broken for a very long time before the fix that changed the "number prefix" from # to $.

There are more changes I would like to make, like supporting default parameters for plugin ports,
avoiding the constant buffer copying if FluidSynth is compiled with float instead of double, etc. But before I do that, I wanted to know show the current state first and ask for feedback.

### CODE REVIEW

**Comment by mawe42** [comment]
@derselbst Would it be possible to get another Travis build with -Denable-ladspa=1?

File: src/bindings/fluid_ladspa.h [hunk 1]
```diff
@@ -21,220 +21,129 @@
...
+    fluid_ladspa_plugin_t *plugins[FLUID_LADSPA_MAX_PLUGINS];
+    int num_plugins;
+
+    /* used to generate the unique plugin ids */
+    int next_plugin_id;
+
+    fluid_rec_mutex_t api_mutex;
 
-#endif  /* _FLUID_LADSPA_H */
+    volatile int state;
```
Comment (derselbst) [inline src/bindings/fluid_ladspa.h]: why volatile?

File: src/rvoice/fluid_rvoice_mixer.c [hunk 2]
```diff
@@ -642,10 +642,20 @@ void delete_fluid_rvoice_mixer(fluid_rvoice_mixer_t* mixer)
...
+  mixer->ladspa_fx = ladspa_fx;
+}
+
+void fluid_rvoice_mixer_deactivate_ladspa(fluid_rvoice_mixer_t *mixer)
+{
+    if (mixer->ladspa_fx == NULL) return;
+    mixer->ladspa_fx->state = FLUID_LADSPA_INACTIVE;
+
+    /* Signal LADSPA engine that deactivation has been processed */
+    fluid_cond_mutex_lock(mixer->ladspa_fx->state_mutex);
```
Comment (derselbst) [inline src/rvoice/fluid_rvoice_mixer.c]: mutex lock in `rvoice_mixer`. Is that a good idea?

File: src/synth/fluid_synth.c [hunk 3]
```diff
@@ -5269,3 +5270,24 @@ int fluid_synth_set_channel_type(fluid_synth_t* synth, int chan, int type)
...
 }
 
+#ifdef LADSPA
+/**
+ * Deactivate the LADSPA effects
+ *
+ * @param synth FluidSynth instance
+ * @return FLUID_OK on success, FLUID_FAILED otherwise
+ */
+int fluid_synth_deactivate_ladspa(fluid_synth_t *synth)
```
Comment (derselbst) [inline src/synth/fluid_synth.c]: So the only purpose is to deactivate ladspa if the user changes his mind? Shall it be part of public API (FLUIDSYNTH_API)?

File: src/bindings/fluid_ladspa.h [hunk 4]
```diff
@@ -21,220 +21,129 @@
...
-int fluid_LADSPA_handle_start(fluid_synth_t* synth, int ac, char** av, fluid_ostream_t out);
+    const LADSPA_Descriptor *desc;
+    LADSPA_Handle *handle;
 
-/* Purpose:
- * Adds one plugin into the list of the LADSPA Fx unit.
- * This is only allowed, while the Fx block is in 'bypass' state (after clear).
- */
-int fluid_LADSPA_handle_add(fluid_synth_t* synth, int ac, char** av, fluid_ostream_t out);
+    volatile int active;
```
Comment (derselbst) [inline src/bindings/fluid_ladspa.h]: why volatile?

File: src/bindings/fluid_ladspa.c [hunk 5]
```diff
@@ -22,1107 +22,1288 @@
...
+    fx->state_mutex = new_fluid_cond_mutex();
+    if (fx->state_mutex == NULL)
+    {
+        FLUID_FREE(fx);
+        return NULL;
+    }
+    fx->state_cond = new_fluid_cond();
+    if (fx->state_cond == NULL)
+    {
+        FLUID_FREE(fx);
```
Comment (derselbst) [inline src/bindings/fluid_ladspa.c]: missing `delete_fluid_cond_mutex(fx->state_mutex);`

File: src/bindings/fluid_ladspa.c [hunk 6]
```diff
@@ -22,1107 +22,1288 @@
...
+
+    fx = FLUID_NEW(fluid_ladspa_fx_t);
+    if (fx == NULL)
+    {
+        return NULL;
+    }
+
+    fx->state = FLUID_LADSPA_INACTIVE;
+
+    fx->sample_rate = synth->sample_rate;
```
Comment (derselbst) [inline src/bindings/fluid_ladspa.c]: I'm not a fan of shadowing variables. And the synth may actually change the samplerate. I guess this would cause trouble. Couldnt you store the synth instance and use the synth API getters instead?

File: src/bindings/fluid_ladspa.c [hunk 7]
```diff
@@ -22,1107 +22,1288 @@
...
+static int create_input_output_nodes(fluid_ladspa_fx_t *fx)
+{
+    char name[99];
+    int i;
+
+    /* These nodes transport the main audio generated by FluidSynth into the
+     * LADSPA effects unit. Create left and right input nodes for each audio group. */
+    for (i = 0; i < fx->audio_groups; i++)
+    {
+        FLUID_SPRINTF(name, "in%i_L", (i + 1));
```
Comment (derselbst) [inline src/bindings/fluid_ladspa.c]: use FLUID_S**N**PRINTF pls

`FLUID_SNPRINTF(name, sizeof(name), "in%i_L", (i + 1));`

File: src/bindings/fluid_ladspa.c [hunk 8]
```diff
@@ -22,1107 +22,1288 @@
...
+ * Does a case insensitive 'fuzzy' prefix match.
+ *
+ * Checks if haystack starts with needle, ignoring case differences and
+ * treating spaces and underscores as equal.
+ *
+ * @param haystack string to search in
+ * @param needle string prefix to search for
+ * @return TRUE if match was found, otherwise FALSE
+ */
+static int fuzzy_prefix_match(const char *haystack, const char *needle)
```
Comment (derselbst) [inline src/bindings/fluid_ladspa.c]: Couldnt you achieve easier that with `FLUID_STRCASECMP` ?

File: src/bindings/fluid_ladspa.c [hunk 9]
```diff
@@ -22,1107 +22,1288 @@
...
-    printf("***Error032***\n"
-	     "Too many plugins.\nChange FLUID_LADSPA_MaxPlugins!");
-    fluid_LADSPA_clear(FxUnit);
-    return(PrintErrorMessage);
-  };
-
-  /* CommandLine (token sequence) is terminated with NULL.
-   * Add two more NULLs, so that a chunk of three tokens can be checked later without risk.*/
-
-  CommandLine=FLUID_ARRAY(char*, (ac+3));assert(CommandL
```
Comment (derselbst) [inline src/bindings/fluid_ladspa.c]: I would prefer just handing it to `delete_fluid_ladspa_plugin()` than doing incremental cleanup.

Comment (derselbst) [inline src/bindings/fluid_ladspa.c]: same here: just call `delete_fluid_ladspa_node()`

**Comment by derselbst** [comment]
will look into travis

File: src/bindings/fluid_ladspa.c (hunk 6 above)
Comment (mawe42) [reply src/bindings/fluid_ladspa.c]: The LADSPA engine can't handle sample rate changes yet, so I thought it would be best to make this more or less obvious by copying that setting on instantiation of the LADSPA fx. But this is only temporary, I will add another commit that introduces proper handling of sample rate changes and will address this comment then. Would that be ok for you?

Comment (mawe42) [reply src/bindings/fluid_ladspa.c]: Just to be clear... the old LADSPA engine couldn't handle sample rate changes either. Even worse... the sample rate was hardcoded to 44100.

File: src/bindings/fluid_ladspa.c (hunk 5 above)
Comment (mawe42) [reply src/bindings/fluid_ladspa.c]: Ack, fixed.

File: src/bindings/fluid_ladspa.c (hunk 7 above)
Comment (mawe42) [reply src/bindings/fluid_ladspa.c]: Ack, fixed

File: src/bindings/fluid_ladspa.c (hunk 8 above)
Comment (mawe42) [reply src/bindings/fluid_ladspa.c]: Yes, the case-insensitive prefix match would be much easier iwth strcasecmp. But there is still the spaces-to-underscores translation, so I would have to copy the string and iterate over it anyway. Might still be cleaner... the current implementation is simply a cleaned up version of the old code.

File: src/bindings/fluid_ladspa.c (hunk 9 above)
Comment (mawe42) [reply src/bindings/fluid_ladspa.c]: Sorry, I don't understand what you are referring to here. The marked code is all deleted.

Comment (mawe42) [reply src/bindings/fluid_ladspa.c]: Same here... :-) What are you referring to?

File: src/bindings/fluid_ladspa.h (hunk 4 above)
Comment (mawe42) [reply src/bindings/fluid_ladspa.h]: To ensure this variable is guaranteed to be updated before the API mutex is released. Otherwise there could be a race condition, leading to a plugin being activated or deactivated more than once (which is forbidden in the LADSPA specs).

File: src/bindings/fluid_ladspa.h (hunk 1 above)
Comment (mawe42) [reply src/bindings/fluid_ladspa.h]: To ensure that the state variable is guaranteed to be written before the API mutex is released. Otherwise there could be a race condition, leading to two command handlers activating or deactivating the effects engine more than once.

File: src/bindings/fluid_ladspa.c (hunk 9 above)
Comment (derselbst) [reply src/bindings/fluid_ladspa.c]: Oh, github messed this up or I made some wrong click.

I referred to your "failed alloc" handling in `new_fluid_ladspa_plugin()`. I would prefer just giving that to the destructor function instead of doing custom cleanups.

Comment (derselbst) [reply src/bindings/fluid_ladspa.c]: again failed alloc handling but for `new_fluid_ladspa_node()`

File: src/rvoice/fluid_rvoice_mixer.c (hunk 2 above)
Comment (mawe42) [reply src/rvoice/fluid_rvoice_mixer.c]: I wondered as well, but the old code did exactly that in fluid_LADSPA_run. I don't think it's a problem, because the mutex is only used for the cond signalling. So it's going to be a wait-free lock in almost 100% of cases. And that code is not called very often, currently only if a user enters ladsap_reset or ladspa_stop into the shell.

File: src/synth/fluid_synth.c (hunk 3 above)
Comment (mawe42) [reply src/synth/fluid_synth.c]: No need for a public API here, this should really be called via fluid_ladspa_deactivate. I think some of the fluid_ladspa_* functions could be made public, thought. But I suggest leaving that for later.

File: doc/ladspa.md [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+    ladspa_plugin delay.so delay_5s
+    ladspa_port Input < in1_L
+    ladspa_port Output > out1_L
+    ladspa_port Delay = 1.0
+    ladspa_port Dry/Wet = 0.5
+
+    ladspa_start
+
+The audible effect should be an untouch right channel and a slightly
```
[the remark this answers is not in the document]
Comment (mawe42) [reply doc/ladspa.md]: Thanks, fixed.

File: doc/ladspa.md [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+binding can be used to route the FluidSynth audio output through any number
+of LADSPA plugins. As the name implies, it is only available on Linux.
+
+## Configuration
+
+To configure and compile FluidSynth with LADSPA support, make sure you have
+the LADSPA SDK (basically the ladspa.h header file) installed. Then enable
+LADSPA when calling cmake:
+
+    cmake path/to/source/ -Denable-ladspa=1
```
[the remark this answers is not in the document]
Comment (mawe42) [reply doc/ladspa.md]: Thanks, fixed.

File: src/bindings/fluid_ladspa.c (hunk 6 above)
Comment (derselbst) [reply src/bindings/fluid_ladspa.c]: I'm a bit two minded about this. On the one hand I would accept it, on the other "Nothing lasts longer than a temporary..."

File: src/rvoice/fluid_rvoice_mixer.c (hunk 2 above)
Comment (derselbst) [reply src/rvoice/fluid_rvoice_mixer.c]: Agreed, ok then.

File: src/bindings/fluid_ladspa.h (hunk 4 above)
Comment (derselbst) [reply src/bindings/fluid_ladspa.h]: You're always accessing it via a structure. I think it should be updated properly. However if you want guarantees you should use atomic setters to be sure.

File: src/bindings/fluid_ladspa.c (hunk 6 above)
Comment (mawe42) [reply src/bindings/fluid_ladspa.c]: You are so right :-) Let me rephrase: I will address this before signing off on this pull request.

File: src/bindings/fluid_ladspa.c (hunk 9 above)
Comment (mawe42) [reply src/bindings/fluid_ladspa.c]: So initialize all pointers that are going to be malloc'ed to NULL just after creation, change the destructor to check for NULL for those pointers and call destructor on each malloc fail?

Comment (derselbst) [reply src/bindings/fluid_ladspa.c]: Yes. `memset(this, 0, sizeof(this));` at the beginning of `new_*()`. And then you can safely call `delete_*()`, see as it has been done in the synth: https://github.com/FluidSynth/fluidsynth/blob/bf3b5a9419080054a9f0fb08f90b4cd7de2f911a/src/synth/fluid_synth.c#L736

Comment (mawe42) [reply src/bindings/fluid_ladspa.c]: Ok, fixed

File: src/bindings/fluid_ladspa.c (hunk 6 above)
Comment (mawe42) [reply src/bindings/fluid_ladspa.c]: ce79b5f adds sample rate change handling. I still shadow the sample rate in the ladspa structure, because the plugins want it in unsigned long and not float. And also because I think it's better to work with a stable value for the sample rate and only change it in fluid_ladspa_set_sample_rate. What do you think?

File: src/bindings/fluid_ladspa.h (hunk 4 above)
Comment (mawe42) [reply src/bindings/fluid_ladspa.h]: Sorry, wrong idea and wrong explanation. What I really want the volatile to protect the code against is weird compiler optimizations. The optimizer might try to store it in a local register and never re-read it again if it's not declared volatile. That might happen on the while loop in fluid_ladspa_deactivate. And as far as I know, accessing a struct member as opposed to a "normal" variable makes no difference to the compiler... but I might be mistaken?

On the plugin->active I think you are right, it is unnecessary and I will remove it.

File: src/bindings/fluid_ladspa.c [hunk 12]
```diff
@@ -160,6 +162,43 @@ void delete_fluid_ladspa_fx(fluid_ladspa_fx_t *fx)
...
+ *
+ * Resets the LADSPA effects if the sample rate is different from the
+ * previous sample rate.
+ *
+ * @param fx LADSPA fx instance
+ * @param sample_rate new sample rate
+ */
+void fluid_ladspa_set_sample_rate(fluid_ladspa_fx_t *fx, fluid_synth_t *synth)
+{
+    unsigned long new_sample_rate = (unsigned long)synth->sample_rate;
```
Comment (derselbst) [inline src/bindings/fluid_ladspa.c]: `synth->sample_rate` is mutex protected, potentially causing a dirty read here. Prefer handing `new_sample_rate` directly into `fluid_ladspa_set_sample_rate()` as extra param.

Also consider doing this:

`(unsigned long)(synth->sample_rate + 0.5) /* to minimize overall casting error */`

File: src/synth/fluid_synth.c [hunk 13]
```diff
@@ -2383,6 +2383,16 @@ fluid_synth_set_sample_rate(fluid_synth_t* synth, float sample_rate)
   fluid_synth_update_mixer(synth, fluid_rvoice_mixer_set_samplerate, 
 			   0, sample_rate);
   fluid_synth_api_exit(synth);
```
Comment (derselbst) [inline src/synth/fluid_synth.c]: I'm afraid we are giving up our lock too early here. Consider this thread gets suspended right after  `fluid_synth_api_exit()` and the next call would be `delete_fluid_synth()` deleting `synth->ladspa_fx`.

On the other hand I'm not sure if we might experience a deadlock waiting to enter LADSPA_API.

Comment (derselbst) [reply src/synth/fluid_synth.c]: And ofc if you're postponing `fluid_synth_api_exit()` so `ladspa_set_sample_rate` is protected you can ignore my first comment.

Comment (mawe42) [reply src/synth/fluid_synth.c]: Yes, it deadlocks if the fluid API is not released before calling fluid_ladspa_set_sample_rate. Hm...

File: src/bindings/fluid_ladspa.h (hunk 4 above)
Comment (derselbst) [reply src/bindings/fluid_ladspa.h]: Ok, then just add a small comment to `state`, pointing out that register optimization thingy.

File: src/synth/fluid_synth.c (hunk 13 above)
Comment (mawe42) [reply src/synth/fluid_synth.c]: Ah... we could handle this via the mixer event queue as well. Then there would be no deadlock. I'll draft the idea in a commit.

File: src/bindings/fluid_ladspa.h (hunk 4 above)
Comment (derselbst) [reply src/bindings/fluid_ladspa.h]: I still think `state` is not optimized to a register, because it's properly accessed via struct and the compiler cannot be sure that this memory isnt aliased, because it doesnt live in local scope.

But since it's performance irrelevant it probably doesnt hurt to be volatile.

File: src/bindings/fluid_ladspa.c (hunk 8 above)
Comment (derselbst) [reply src/bindings/fluid_ladspa.c]: We can also ignore that for now, so you can concentrate on the samplerate change.

**Comment by mawe42** [comment]
Ok, I've reworked a lot of the LADSPA engine. It is now completely decoupled from FluidSynth. Everything it needs to function (sample rate, number  of input and output channels) is passed in on creation. Beside the obvious benefit of easier maintenance of FluidSynth and the LADSPA engine due to the loose coupling, it also has an additional benefit in that users can create their own (even additional) effect engines for further or other processing. After we expose a public API for LADSPA, of course.

Synchronization is done only with atomic operations now. And the sample rate change trigger is also greatly simplified now, only a very small hook in FluidSynths fluid_rvoice_mixer_set_samplerate.

I'm quite happy with it now. The only snag is the g_usleep call, which isn't available as a FLUID_* macro yet. But that can easily be added. Also, while explicit sleep calls like this might seem dirty, it is only used as a fallback. During my tests of starting and stopping LADSPA while FluidSynth was rendering sound, I never managed to actually trigger this sleep.

Looking forward to your feedback! Thanks for taking the time to get this change polished!

**Comment by derselbst** [comment]
That timeout strikes me. Apart from that it's absolutely fine.

**Comment by derselbst** [comment]
Anyway, I think I'll just merge it and look at it later again.

File: src/bindings/fluid_ladspa.c (hunk 8 above)
Comment (mawe42) [reply src/bindings/fluid_ladspa.c]: I've removed the weird fuzzy_prefix_match function. It's not needed anymore anyway, as the command line parser knows how to deal with quoted names like "Dry / Wet Mix". They are passed in as one string. Needed to add FLUID_STRNCASECMP to fluid_priv.h though.

**Comment by derselbst** [comment]
cecf115a38f35e1c25be5b0f2eb9ef67827616de: Oh ok, nice. Waiting for you being ready.

**Comment by mawe42** [comment]
> That timeout strikes me. Apart from that it's absolutely fine.

Ok, then dont' merge yet, please. It could probably also be solved with a cond and mutex...

**Comment by derselbst** [comment]
> cond and mutex...

Yes, that would be my favourite.

**Comment by mawe42** [comment]
Done, and much nicer than the timeout. Thanks for insisting :-)

**Comment by derselbst** [comment]
Yeah, that looks good. Thanks for fixing, really nice work! Will merge it once travis done.

**Comment by mawe42** [comment]
Added one last cleanup commit, removing unused includes. I will leave this branch alone now.
